1. Design model 
    - Input size
    - Output size
    - Forward Pass
1. Construct Loss and Optimizer
1. Training Loop
    1. Forward Pass: Compute Prediction
    1. Backward Pass: Gradients
    1. Update Weights

In [1]:
import torch
import torch.nn as nn

In [2]:
# now as we are using model from torch
# x and y need to be 2 dimentional
# number of rows -> number of samples
# each row -> features
x = torch.tensor([[1], [2], [3], [4]], dtype=torch.float32)
y = torch.tensor([[2], [4], [6], [8]], dtype=torch.float32)

nSamples, nFeatures = x.shape
print(f"Number of samples: {nSamples}")
print(f"Number of features: {nFeatures}")

Number of samples: 4
Number of features: 1


In [3]:
xTest = torch.tensor([5], dtype=torch.float32)

In [4]:
#Model
inputSize = nFeatures
outputSize = nFeatures

model = nn.Linear(inputSize, outputSize)

In [5]:
learning_rate = 0.01
n_iters = 100

loss = nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

In [6]:
for epoch in range(n_iters):
    #prediction = foward pass
    y_pred = model(x)

    #loss
    l = loss(y, y_pred)

    #gradients = backward pass
    l.backward()
    
    #update weights
        # with torch.no_grad():
        #     w -= learning_rate * w.grad
        # we don't need to manually update weights here
    optimizer.step()

    # zero gradients
        # w.grad.zero_()
    optimizer.zero_grad()

    if epoch % 10 == 0:
        [w, b] = model.parameters()
        print(f"epoch {epoch+1}: w = {w[0][0].item():.3f}, loss = {l:.8f}")

print(f"Prediction after training: f(5) = {model(xTest).item():.3f}")        
    

epoch 1: w = 0.120, loss = 25.46641159
epoch 11: w = 1.282, loss = 1.00921631
epoch 21: w = 1.481, loss = 0.35605887
epoch 31: w = 1.525, loss = 0.31995535
epoch 41: w = 1.544, loss = 0.30093449
epoch 51: w = 1.558, loss = 0.28340822
epoch 61: w = 1.571, loss = 0.26691201
epoch 71: w = 1.584, loss = 0.25137645
epoch 81: w = 1.596, loss = 0.23674494
epoch 91: w = 1.608, loss = 0.22296520
Prediction after training: f(5) = 9.214


as the model is randomly initializing weights and biases<br> evertime you run the code<br> you get different and maybe sometimes unsatisfying results<br>
sometimes you'll get w closer to 2 in 30 epoches<br> other times you'll get w as 1.5 even after 100 epoches